sdks/python/apache_beam/runners/direct/executor.py

```diff
@@ -545,22 +545,19 @@ def call(self):
           self._executor.executor_service.submit(self)
 
     def _should_shutdown(self):
-      """_should_shutdown checks whether pipeline is completed or not.
+      """Checks whether the pipeline is completed and should be shut down.
 
-      It will check for successful completion by checking the watermarks of all
-      transforms. If they all reached the maximum watermark it means that
-      pipeline successfully reached to completion.
+      If there is anything in the queue of tasks to do, do not shut down.
```

Comment:
<!--thread_id:cc_153932412_t; commit:517fae9dcb749a59322ea40c7de0eb90ebf5efac; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Are we checking whether there are pending processing time timers?

<!--section:body-->
Yes, and I consider that a "task to do," which is looked at in _is_executing() and that is where I added a comment about timers being set.